File: src/next.test.ts
import { describe, expect, test } from 'bun:test';
import { formatNextCandidates, NEXT_DEFAULT_LIMIT, rankNext, toJsonCandidates } from './next';
import type { TodoView } from './refs';

/** 기준 시각 — 2026-07-30 09:00 로컬. 마감 라벨은 이 날짜를 "오늘" 로 읽는다. */
const NOW = new Date(2026, 6, 30, 9, 0, 0).getTime();

let seq = 0;

function todo(overrides: Partial<TodoView> = {}): TodoView {
  seq += 1;
  return {
    id: `t${seq}`,
    number: seq,
    boardId: 'b1',
    title: `할 일 ${seq}`,
    description: '',
    status: 'todo',
    priority: 'p4',
    labels: [],
    links: [],
    position: seq,
    createdAt: '2026-07-01T00:00:00.000Z',
    updatedAt: '2026-07-01T00:00:00.000Z',
    ref: `rocky-${seq}`,
    commentCount: 0,
    ...overrides,
  };
}

function refs(candidates: ReturnType<typeof rankNext>): string[] {
  return candidates.map((c) => c.todo.ref);
}

describe('rankNext — 후보에서 빠지는 것', () => {
  test('done 과 보관된 항목은 후보가 아니다', () => {
    const open = todo();
    const ranked = rankNext(
      [open, todo({ status: 'done' }), todo({ archivedAt: '2026-07-20T00:00:00.000Z' })],
      { now: NOW },
    );
    expect(refs(ranked)).toEqual([open.ref]);
  });

  test('살아 있는 세션이 붙들고 있는 doing 은 후보가 아니다', () => {
    const open = todo();
    const held = todo({ status: 'doing', doingBy: 'claude-code', doingState: 'live' });
    expect(refs(rankNext([open, held], { now: NOW }))).toEqual([open.ref]);
  });

  test('열린 자식을 가진 부모(우산 항목)는 빠지고 자식이 남는다', () => {
    const parent = todo({ priority: 'p1' });
    const child = todo({ parentId: parent.id });
    expect(refs(rankNext([parent, child], { now: NOW }))).toEqual([child.ref]);
  });

  test('자식이 전부 done 인 부모는 남는다 — 마무리가 남았다', () => {
    const parent = todo();
    const child = todo({ parentId: parent.id, status: 'done' });
    expect(refs(rankNext([parent, child], { now: NOW }))).toEqual([parent.ref]);
  });
});

describe('rankNext — 순서', () => {
  test('주인 없는 doing 이 p1 보다 위다', () => {
    const urgent = todo({ priority: 'p1' });
    const orphan = todo({ status: 'doing', doingBy: 'claude-code', doingState: 'gone' });
    expect(refs(rankNext([urgent, orphan], { now: NOW }))).toEqual([orphan.ref, urgent.ref]);
  });

  test('세션 없음(gone)이 멈춤(idle)보다 위다', () => {
    const idle = todo({ status: 'doing', doingBy: 'claude-code', doingState: 'idle' });
    const gone = todo({ status: 'doing', doingBy: 'claude-code', doingState: 'gone' });
    expect(refs(rankNext([idle, gone], { now: NOW }))).toEqual([gone.ref, idle.ref]);
  });

  test('주인 없는 doing 은 하위 범주가 다 쌓여도 밀리지 않는다', () => {
    // 리뷰 반례: 합산 점수였을 때 `gone` p4(100) 가 마감 지난 p1 + 최근 댓글(112) 에 밀렸다.
    const orphan = todo({ priority: 'p4', status: 'doing', doingState: 'gone' });
    const loaded = todo({
      priority: 'p1',
      due: '2026-07-20',
      lastCommentAt: '2026-07-30T08:00:00.000Z',
    });
    expect(refs(rankNext([loaded, orphan], { now: NOW }))).toEqual([orphan.ref, loaded.ref]);
  });

  test('마감은 우선순위·최근 댓글이 다 쌓여도 이긴다', () => {
    const due = todo({ priority: 'p4', due: '2026-08-05' });
    const loaded = todo({ priority: 'p1', lastCommentAt: '2026-07-30T08:00:00.000Z' });
    expect(refs(rankNext([loaded, due], { now: NOW }))).toEqual([due.ref, loaded.ref]);
  });

  test('판정 불가한 진행중은 마감 아래, 우선순위 위다', () => {
    const due = todo({ priority: 'p4', due: '2026-08-05' });
    const inProgress = todo({ priority: 'p4', status: 'doing', doingBy: 'logan' });
    const urgent = todo({ priority: 'p1' });
    expect(refs(rankNext([urgent, inProgress, due], { now: NOW }))).toEqual([
      due.ref,
      inProgress.ref,
      urgent.ref,
    ]);
  });

  test('마감이 지난 항목이 p1 보다 위다', () => {
    const urgent = todo({ priority: 'p1' });
    const overdue = todo({ priority: 'p4', due: '2026-07-28' });
    expect(refs(rankNext([urgent, overdue], { now: NOW }))).toEqual([overdue.ref, urgent.ref]);
  });

  test('우선순위 내림차순 — p1 → p2 → p3 → p4', () => {
    const p4 = todo({ priority: 'p4' });
    const p2 = todo({ priority: 'p2' });
    const p1 = todo({ priority: 'p1' });
    const p3 = todo({ priority: 'p3' });
    expect(refs(rankNext([p4, p2, p1, p3], { now: NOW }))).toEqual([
      p1.ref,
      p2.ref,
      p3.ref,
      p4.ref,
    ]);
  });

  test('동점은 position → number 로 결정적으로 갈린다', () => {
    const later = todo({ position: 9 });
    const earlier = todo({ position: 2 });
    expect(refs(rankNext([later, earlier], { now: NOW }))).toEqual([earlier.ref, later.ref]);
  });

  test('limit 은 상위 N 개만 남긴다', () => {
    const items = [todo({ priority: 'p1' }), todo({ priority: 'p2' }), todo({ priority: 'p3' })];
    expect(rankNext(items, { now: NOW, limit: 2 })).toHaveLength(2);
    expect(rankNext(items, { now: NOW })).toHaveLength(3);
  });
});

describe('rankNext — 근거 라벨', () => {
  test('마감은 D-day / D-n / D+n 으로 읽힌다', () => {
    const reasonOf = (due: string) => rankNext([todo({ due })], { now: NOW })[0]?.reason;
    expect(reasonOf('2026-07-30')).toBe('마감 D-day');
    expect(reasonOf('2026-08-02')).toBe('마감 D-3');
    expect(reasonOf('2026-07-27')).toBe('마감 D+3');
  });

  test('마감이 멀면 라벨이 붙지 않는다', () => {
    expect(rankNext([todo({ due: '2026-12-25' })], { now: NOW })[0]?.reason).toBe('대기 중');
  });

  test('p3/p4 는 우선순위를 근거로 찍지 않는다', () => {
    expect(rankNext([todo({ priority: 'p2' })], { now: NOW })[0]?.reason).toBe('p2');
    expect(rankNext([todo({ priority: 'p3' })], { now: NOW })[0]?.reason).toBe('대기 중');
  });

  test('근거는 강한 것부터 이어 붙는다', () => {
    const item = todo({
      status: 'doing',
      doingBy: 'claude-code',
      doingState: 'idle',
      priority: 'p2',
      due: '2026-07-31',
      lastCommentAt: '2026-07-30T08:00:00.000Z',
    });
    expect(rankNext([item], { now: NOW })[0]?.reason).toBe(
      '이어받기(멈춤) · 마감 D-1 · p2 · 최근 댓글',
    );
  });

  test('오래된 댓글은 근거가 아니다', () => {
    const stale = todo({ lastCommentAt: '2026-07-01T00:00:00.000Z' });
    expect(rankNext([stale], { now: NOW })[0]?.reason).toBe('대기 중');
  });

  test('판정 불가한 doing 은 이어받기라고 부르지 않는다', () => {
    const human = todo({ status: 'doing', doingBy: 'logan', doingState: 'unknown' });
    expect(rankNext([human], { now: NOW })[0]?.reason).toBe('진행중(logan)');
  });
});

describe('rankNext — 깨진 입력', () => {
  test('망가진 due 는 0점으로 흘리고 던지지 않는다', () => {
    const broken = todo({ due: '2026-6' });
    const ranked = rankNext([broken], { now: NOW });
    expect(ranked).toHaveLength(1);
    expect(ranked[0]?.reason).toBe('대기 중');
  });

  test('달력에 없는 날짜는 마감으로 세지 않는다 — Date.UTC 가 굴려 버리는 값들', () => {
    // due 는 어디서도 검증되지 않는 자유 문자열이라 실제로 저장될 수 있다.
    // 2026-02-31 → 3월 3일, 2026-13-01 → 2027년 1월 1일로 굴러가 엉뚱한 D-day 가 찍혔다.
    const reasonOf = (due: string) => rankNext([todo({ due })], { now: NOW })[0]?.reason;
    expect(reasonOf('2026-02-31')).toBe('대기 중');
    expect(reasonOf('2026-13-01')).toBe('대기 중');
    expect(reasonOf('2026-00-10')).toBe('대기 중');
    // 경계: 실재하는 날짜(윤년 2월 29일 포함)는 그대로 마감으로 세어져야 한다.
    expect(reasonOf('2026-02-28')).toContain('마감 D+');
    expect(reasonOf('2024-02-29')).toContain('마감 D+');
  });

  test('마감 지남 판정은 실재하는 날짜에서만 나온다', () => {
    expect(rankNext([todo({ due: '2026-07-31' })], { now: NOW })[0]?.reason).toBe('마감 D-1');
    // 같은 자리수인데 달력에 없는 값 — 위와 달리 아무 라벨도 붙지 않아야 한다.
    expect(rankNext([todo({ due: '2026-07-32' })], { now: NOW })[0]?.reason).toBe('대기 중');
  });

  test('빈 보드는 빈 배열', () => {
    expect(rankNext([], { now: NOW })).toEqual([]);
  });
});

describe('toJsonCandidates — payload 는 컴팩트해야 한다', () => {
  const keyOf = (id: string) => (id === 'b1' ? 'rocky-todo' : undefined);

  test('TodoView 전체가 아니라 고를 때 필요한 필드만 나간다', () => {
    const item = todo({ title: '오리진 검사', priority: 'p2', description: '본문' });
    const [row] = toJsonCandidates(rankNext([item], { now: NOW }), keyOf);
    expect(row).toEqual({
      ref: item.ref,
      number: item.number,
      board: 'rocky-todo',
      title: '오리진 검사',
      reason: 'p2',
      priority: 'p2',
      status: 'todo',
      due: undefined,
      labels: [],
      commentCount: 0,
      summary: '본문',
    });
    // 커맨드가 쓰지 않는 무거운 필드는 실리지 않는다 — 이게 응답 지연의 원인이었다.
    expect(row).not.toHaveProperty('description');
    expect(row).not.toHaveProperty('links');
    expect(row).not.toHaveProperty('doingSessionId');
  });

  test('summary 는 여러 줄을 한 줄로 눌러 160자까지 자른다', () => {
    const long = todo({ description: `첫 줄\n\n${'가'.repeat(300)}` });
    const [row] = toJsonCandidates(rankNext([long], { now: NOW }), keyOf);
    expect(row?.summary).toStartWith('첫 줄 가가가');
    expect(row?.summary).toEndWith('…');
    expect(row?.summary?.length).toBe(161); // 160자 + 말줄임표
  });

  test('description 이 비면 summary 필드 자체가 없다', () => {
    const [row] = toJsonCandidates(rankNext([todo({ description: '' })], { now: NOW }), keyOf);
    expect(row?.summary).toBeUndefined();
  });

  test('board key 를 못 찾아도 던지지 않고 빈 문자열로 둔다', () => {
    const orphanBoard = todo({ boardId: 'unknown' });
    const [row] = toJsonCandidates(rankNext([orphanBoard], { now: NOW }), keyOf);
    expect(row?.board).toBe('');
  });
});

describe('formatNextCandidates', () => {
  test('한 줄에 ref · 제목 · 근거', () => {
    const ranked = rankNext([todo({ title: '오리진 검사', priority: 'p2' })], { now: NOW });
    expect(formatNextCandidates(ranked)).toBe(`1. ${ranked[0]?.todo.ref}  오리진 검사  — p2`);
  });

  test('개행이 든 제목도 한 줄로 눌린다 — 한 줄 = 후보 하나', () => {
    // 제목은 어디서도 다듬어지지 않는다(REST 는 빈 문자열만 거부, MCP 는 z.string()).
    // 그대로 보간하면 후보 하나가 여러 줄로 쪼개져 사용자가 고른 번호와 ref 가 어긋난다.
    const nasty = todo({ title: '앞줄\n2. 가짜 후보\t뒤줄' });
    const out = formatNextCandidates(rankNext([nasty], { now: NOW }));
    expect(out.split('\n')).toHaveLength(1);
    expect(out).toContain('앞줄 2. 가짜 후보 뒤줄');
  });

  test('후보가 없으면 이유를 말한다', () => {
    expect(formatNextCandidates([])).toContain('후보가 없다');
  });

  test('기본 후보 수는 한 화면에서 훑을 만큼이다', () => {
    expect(NEXT_DEFAULT_LIMIT).toBe(8);
  });
});

File: crates/rocky-todo-core/tests/next_test.rs
//! TS 원본 `src/next.test.ts` 포팅.

use rocky_todo_core::next::{
    format_next_candidates, rank_next, to_json_candidates, NextCandidate, RankNextOptions,
    NEXT_DEFAULT_LIMIT,
};
use rocky_todo_core::refs::TodoView;
use rocky_todo_core::types::*;
use std::sync::atomic::{AtomicI64, Ordering};

/// 기준 시각 — 2026-07-30 09:00 로컬.
fn now_ms() -> i64 {
    use chrono::TimeZone;
    chrono::Local
        .with_ymd_and_hms(2026, 7, 30, 9, 0, 0)
        .single()
        .unwrap()
        .timestamp_millis()
}

static SEQ: AtomicI64 = AtomicI64::new(0);

fn todo() -> TodoView {
    let seq = SEQ.fetch_add(1, Ordering::Relaxed) + 1;
    TodoView {
        todo: Todo {
            id: format!("t{seq}"),
            number: seq,
            board_id: "b1".into(),
            section_id: None,
            parent_id: None,
            title: format!("할 일 {seq}"),
            description: String::new(),
            status: TodoStatus::Todo,
            priority: TodoPriority::P4,
            due: None,
            labels: vec![],
            links: vec![],
            doing_by: None,
            doing_since: None,
            doing_session_id: None,
            position: seq,
            created_at: "2026-07-01T00:00:00.000Z".into(),
            updated_at: "2026-07-01T00:00:00.000Z".into(),
            completed_at: None,
            archived_at: None,
        },
        r#ref: format!("rocky-{seq}"),
        comment_count: 0,
        last_comment_at: None,
        doing_state: None,
    }
}

fn rank(todos: &[TodoView], limit: Option<usize>) -> Vec<NextCandidate> {
    rank_next(
        todos,
        &RankNextOptions {
            now: now_ms(),
            limit,
        },
    )
}

fn refs(candidates: &[NextCandidate]) -> Vec<String> {
    candidates.iter().map(|c| c.todo.r#ref.clone()).collect()
}

fn with_due(mut t: TodoView, due: &str) -> TodoView {
    t.todo.due = Some(due.into());
    t
}

fn reason_of(t: TodoView) -> String {
    rank(&[t], None)[0].reason.clone()
}

use rocky_todo_core::doing::DoingState;

#[test]
fn done_and_archived_are_excluded() {
    let open = todo();
    let mut done = todo();
    done.todo.status = TodoStatus::Done;
    let mut archived = todo();
    archived.todo.archived_at = Some("2026-07-20T00:00:00.000Z".into());
    assert_eq!(
        refs(&rank(&[open.clone(), done, archived], None)),
        vec![open.r#ref]
    );
}

#[test]
fn live_doing_is_excluded() {
    let open = todo();
    let mut held = todo();
    held.todo.status = TodoStatus::Doing;
    held.todo.doing_by = Some("claude-code".into());
    held.doing_state = Some(DoingState::Live);
    assert_eq!(refs(&rank(&[open.clone(), held], None)), vec![open.r#ref]);
}

#[test]
fn umbrella_parent_with_open_child_is_excluded() {
    let mut parent = todo();
    parent.todo.priority = TodoPriority::P1;
    let mut child = todo();
    child.todo.parent_id = Some(parent.todo.id.clone());
    assert_eq!(
        refs(&rank(&[parent, child.clone()], None)),
        vec![child.r#ref]
    );
}

#[test]
fn parent_with_all_done_children_stays() {
    let parent = todo();
    let mut child = todo();
    child.todo.parent_id = Some(parent.todo.id.clone());
    child.todo.status = TodoStatus::Done;
    assert_eq!(
        refs(&rank(&[parent.clone(), child], None)),
        vec![parent.r#ref]
    );
}

#[test]
fn orphan_doing_beats_p1() {
    let mut urgent = todo();
    urgent.todo.priority = TodoPriority::P1;
    let mut orphan = todo();
    orphan.todo.status = TodoStatus::Doing;
    orphan.todo.doing_by = Some("claude-code".into());
    orphan.doing_state = Some(DoingState::Gone);
    assert_eq!(
        refs(&rank(&[urgent.clone(), orphan.clone()], None)),
        vec![orphan.r#ref, urgent.r#ref]
    );
}

#[test]
fn gone_beats_idle() {
    let mut idle = todo();
    idle.todo.status = TodoStatus::Doing;
    idle.todo.doing_by = Some("claude-code".into());
    idle.doing_state = Some(DoingState::Idle);
    let mut gone = todo();
    gone.todo.status = TodoStatus::Doing;
    gone.todo.doing_by = Some("claude-code".into());
    gone.doing_state = Some(DoingState::Gone);
    assert_eq!(
        refs(&rank(&[idle.clone(), gone.clone()], None)),
        vec![gone.r#ref, idle.r#ref]
    );
}

#[test]
fn orphan_is_not_overtaken_by_stacked_lower_bands() {
    // 리뷰 반례: 합산 점수였을 때 gone p4 가 마감 지난 p1 + 최근 댓글에 밀렸다.
    let mut orphan = todo();
    orphan.todo.status = TodoStatus::Doing;
    orphan.doing_state = Some(DoingState::Gone);
    let mut loaded = todo();
    loaded.todo.priority = TodoPriority::P1;
    loaded.todo.due = Some("2026-07-20".into());
    loaded.last_comment_at = Some("2026-07-30T08:00:00.000Z".into());
    assert_eq!(
        refs(&rank(&[loaded.clone(), orphan.clone()], None)),
        vec![orphan.r#ref, loaded.r#ref]
    );
}

#[test]
fn due_beats_stacked_priority_and_comments() {
    let due = with_due(todo(), "2026-08-05");
    let mut loaded = todo();
    loaded.todo.priority = TodoPriority::P1;
    loaded.last_comment_at = Some("2026-07-30T08:00:00.000Z".into());
    assert_eq!(
        refs(&rank(&[loaded.clone(), due.clone()], None)),
        vec![due.r#ref, loaded.r#ref]
    );
}

#[test]
fn unjudged_doing_below_due_above_priority() {
    let due = with_due(todo(), "2026-08-05");
    let mut in_progress = todo();
    in_progress.todo.status = TodoStatus::Doing;
    in_progress.todo.doing_by = Some("logan".into());
    let mut urgent = todo();
    urgent.todo.priority = TodoPriority::P1;
    assert_eq!(
        refs(&rank(
            &[urgent.clone(), in_progress.clone(), due.clone()],
            None
        )),
        vec![due.r#ref, in_progress.r#ref, urgent.r#ref]
    );
}

#[test]
fn overdue_beats_p1() {
    let mut urgent = todo();
    urgent.todo.priority = TodoPriority::P1;
    let overdue = with_due(todo(), "2026-07-28");
    assert_eq!(
        refs(&rank(&[urgent.clone(), overdue.clone()], None)),
        vec![overdue.r#ref, urgent.r#ref]
    );
}

#[test]
fn priority_descending() {
    let mut p4 = todo();
    p4.todo.priority = TodoPriority::P4;
    let mut p2 = todo();
    p2.todo.priority = TodoPriority::P2;
    let mut p1 = todo();
    p1.todo.priority = TodoPriority::P1;
    let mut p3 = todo();
    p3.todo.priority = TodoPriority::P3;
    assert_eq!(
        refs(&rank(
            &[p4.clone(), p2.clone(), p1.clone(), p3.clone()],
            None
        )),
        vec![p1.r#ref, p2.r#ref, p3.r#ref, p4.r#ref]
    );
}

#[test]
fn ties_break_by_position_then_number() {
    let mut later = todo();
    later.todo.position = 9;
    let mut earlier = todo();
    earlier.todo.position = 2;
    assert_eq!(
        refs(&rank(&[later.clone(), earlier.clone()], None)),
        vec![earlier.r#ref, later.r#ref]
    );
}

#[test]
fn limit_keeps_top_n() {
    let mut a = todo();
    a.todo.priority = TodoPriority::P1;
    let mut b = todo();
    b.todo.priority = TodoPriority::P2;
    let mut c = todo();
    c.todo.priority = TodoPriority::P3;
    let items = vec![a, b, c];
    assert_eq!(rank(&items, Some(2)).len(), 2);
    assert_eq!(rank(&items, None).len(), 3);
}

#[test]
fn due_labels_read_as_dday() {
    assert_eq!(reason_of(with_due(todo(), "2026-07-30")), "마감 D-day");
    assert_eq!(reason_of(with_due(todo(), "2026-08-02")), "마감 D-3");
    assert_eq!(reason_of(with_due(todo(), "2026-07-27")), "마감 D+3");
}

#[test]
fn far_due_gets_no_label() {
    assert_eq!(reason_of(with_due(todo(), "2026-12-25")), "대기 중");
}

#[test]
fn p3_p4_get_no_priority_label() {
    let mut p2 = todo();
    p2.todo.priority = TodoPriority::P2;
    assert_eq!(reason_of(p2), "p2");
    let mut p3 = todo();
    p3.todo.priority = TodoPriority::P3;
    assert_eq!(reason_of(p3), "대기 중");
}

#[test]
fn reasons_join_strongest_first() {
    let mut item = todo();
    item.todo.status = TodoStatus::Doing;
    item.todo.doing_by = Some("claude-code".into());
    item.doing_state = Some(DoingState::Idle);
    item.todo.priority = TodoPriority::P2;
    item.todo.due = Some("2026-07-31".into());
    item.last_comment_at = Some("2026-07-30T08:00:00.000Z".into());
    assert_eq!(
        reason_of(item),
        "이어받기(멈춤) · 마감 D-1 · p2 · 최근 댓글"
    );
}

#[test]
fn stale_comment_is_no_reason() {
    let mut stale = todo();
    stale.last_comment_at = Some("2026-07-01T00:00:00.000Z".into());
    assert_eq!(reason_of(stale), "대기 중");
}

#[test]
fn unjudged_doing_is_not_called_takeover() {
    let mut human = todo();
    human.todo.status = TodoStatus::Doing;
    human.todo.doing_by = Some("logan".into());
    human.doing_state = Some(DoingState::Unknown);
    assert_eq!(reason_of(human), "진행중(logan)");
}

#[test]
fn broken_due_flows_as_zero_without_panicking() {
    let ranked = rank(&[with_due(todo(), "2026-6")], None);
    assert_eq!(ranked.len(), 1);
    assert_eq!(ranked[0].reason, "대기 중");
}

#[test]
fn nonexistent_calendar_dates_are_not_due() {
    assert_eq!(reason_of(with_due(todo(), "2026-02-31")), "대기 중");
    assert_eq!(reason_of(with_due(todo(), "2026-13-01")), "대기 중");
    assert_eq!(reason_of(with_due(todo(), "2026-00-10")), "대기 중");
    // 경계: 실재 날짜(윤년 포함)는 마감으로 세어진다.
    assert!(reason_of(with_due(todo(), "2026-02-28")).contains("마감 D+"));
    assert!(reason_of(with_due(todo(), "2024-02-29")).contains("마감 D+"));
}

#[test]
fn overdue_only_from_real_dates() {
    assert_eq!(reason_of(with_due(todo(), "2026-07-31")), "마감 D-1");
    assert_eq!(reason_of(with_due(todo(), "2026-07-32")), "대기 중");
}

#[test]
fn empty_board_is_empty() {
    assert!(rank(&[], None).is_empty());
}

// ── toJsonCandidates ──

fn key_of(id: &str) -> Option<String> {
    if id == "b1" {
        Some("rocky-todo".to_string())
    } else {
        None
    }
}

#[test]
fn json_carries_only_selection_fields() {
    let mut item = todo();
    item.todo.title = "오리진 검사".into();
    item.todo.priority = TodoPriority::P2;
    item.todo.description = "본문".into();
    let expected_ref = item.r#ref.clone();
    let expected_number = item.todo.number;
    let rows = to_json_candidates(&rank(&[item], None), key_of);
    let json = serde_json::to_value(&rows[0]).unwrap();
    assert_eq!(
        json,
        serde_json::json!({
            "ref": expected_ref,
            "number": expected_number,
            "board": "rocky-todo",
            "title": "오리진 검사",
            "reason": "p2",
            "priority": "p2",
            "status": "todo",
            "labels": [],
            "commentCount": 0,
            "summary": "본문",
        })
    );
    assert!(json.get("description").is_none());
    assert!(json.get("links").is_none());
    assert!(json.get("doingSessionId").is_none());
}

#[test]
fn summary_flattens_and_truncates_to_160() {
    let mut long = todo();
    long.todo.description = format!("첫 줄\n\n{}", "가".repeat(300));
    let rows = to_json_candidates(&rank(&[long], None), key_of);
    let summary = rows[0].summary.as_ref().unwrap();
    assert!(summary.starts_with("첫 줄 가가가"));
    assert!(summary.ends_with('…'));
    assert_eq!(summary.chars().count(), 161); // 160자 + 말줄임표
}

#[test]
fn empty_description_omits_summary() {
    let rows = to_json_candidates(&rank(&[todo()], None), key_of);
    assert!(rows[0].summary.is_none());
    let json = serde_json::to_value(&rows[0]).unwrap();
    assert!(json.get("summary").is_none());
}

#[test]
fn missing_board_key_falls_back_to_empty_string() {
    let mut orphan_board = todo();
    orphan_board.todo.board_id = "unknown".into();
    let rows = to_json_candidates(&rank(&[orphan_board], None), key_of);
    assert_eq!(rows[0].board, "");
}

// ── formatNextCandidates ──

#[test]
fn one_line_per_candidate() {
    let mut item = todo();
    item.todo.title = "오리진 검사".into();
    item.todo.priority = TodoPriority::P2;
    let ranked = rank(&[item], None);
    assert_eq!(
        format_next_candidates(&ranked),
        format!("1. {}  오리진 검사  — p2", ranked[0].todo.r#ref)
    );
}

#[test]
fn newlines_in_title_are_flattened() {
    let mut nasty = todo();
    nasty.todo.title = "앞줄\n2. 가짜 후보\t뒤줄".into();
    let out = format_next_candidates(&rank(&[nasty], None));
    assert_eq!(out.split('\n').count(), 1);
    assert!(out.contains("앞줄 2. 가짜 후보 뒤줄"));
}

#[test]
fn empty_candidates_explain_why() {
    assert!(format_next_candidates(&[]).contains("후보가 없다"));
}

#[test]
fn default_limit_fits_one_screen() {
    assert_eq!(NEXT_DEFAULT_LIMIT, 8);
}
